python/lsst/daf/butler/formatters/logs.py

```diff
@@ -0,0 +1,121 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("ButlerLogRecordsFormatter", )
+
+from typing import (
+    Any,
+    Optional,
+    Type,
+)
+
+from lsst.daf.butler import ButlerLogRecords, ButlerLogRecord
+from .json import JsonFormatter
+
+
+class ButlerLogRecordsFormatter(JsonFormatter):
+    """Read and write log records in JSON format.
+
+    This is a naive implementation that treats everything as a pydantic.
+    model.  In the future this may be changed to be able to read
+    ``ButlerLogRecord`` one at time from the file and return a subset
+    of records given some filtering parameters.
+    """
+
+    def _readFile(self, path: str, pytype: Optional[Type[Any]] = None) -> Any:
+        """Read a file from the path in JSON format.
+
+        Parameters
+        ----------
+        path : `str`
+            Path to use to open JSON format file.
+        pytype : `class`, optional
+            Not used by this implementation.
+
+        Returns
+        -------
+        data : `object`
+            Data as Python object read from JSON file.
+
+        Notes
+        -----
+        Can read two forms of JSON log file. It can read a full JSON
+        document created from `ButlerLogRecords`, or a stream off standalone
+        JSON documents with a log record per line.
+        """
+        if pytype is None:
+            pytype = ButlerLogRecords
+        elif not issubclass(pytype, ButlerLogRecords):
+            raise RuntimeError(f"Python type {pytype} does not seem to be a ButlerLogRecords type")
+
+        return pytype.from_file(path)
+
+    def _fromBytes(self, serializedDataset: bytes, pytype: Optional[Type[Any]] = None) -> Any:
+        """Read the bytes object as a python object.
+
+        Parameters
+        ----------
+        serializedDataset : `bytes`
+            Bytes object to unserialize.
+        pytype : `class`, optional
+            Not used by this implementation.
+
+        Returns
+        -------
+        inMemoryDataset : `object`
+            The requested data as a Python object or None if the string could
+            not be read.
+        """
+        # Duplicates some of the logic from ButlerLogRecords.from_file
+        if pytype is None:
+            pytype = ButlerLogRecords
+        elif not issubclass(pytype, ButlerLogRecords):
+            raise RuntimeError(f"Python type {pytype} does not seem to be a ButlerLogRecords type")
+
+        if serializedDataset.startswith(b"["):
+            return pytype.parse_raw(serializedDataset)
+
+        if not serializedDataset.startswith(b"{"):
+            raise RuntimeError(f"These bytes do not look like JSON -- starts with '{serializedDataset[0]}'")
+
+        # Filter out blank lines.
+        records = [ButlerLogRecord.parse_raw(line) for line in serializedDataset.split(b"\n") if line]
```

Comment:
`split()` is probably not super-efficient space-wise for large data volumes.

Reply:
Is there a version of `.split()` somewhere that returns an iterator rather than a list? Or should I walk through the string myself looking for `\n` ?  I think this being a list means that we have all the bytes in memory, and all of them again in a list and all of them accumulating into the output list... On the plus size Datastore has an internal limit as to how big a dataset is allowed to be before using fromBytes.

Reply:
The internet is full of people asking for `str.isplit()` and it never happening.

Reply:
I don't think there is an iterator version of split() in the standard library, but writing a simple generator method for it would be trivial.